core/domain/classifier_services.py

```diff
@@ -0,0 +1,534 @@
+        a new id to the label and returns it.
+        """
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_label_name(self, l):
+        """
+        Returns a label's string name given its internal id.
+        If the id does not have a corresponding name, an exception is raised.
+        """
+        for label_name, label_id in self._label_to_id.iteritems():
+            if label_id == l:
+                return label_name
+
+        raise Exception('Label id %d does not exist.' % l)
+
+    def _get_doc_with_label_vector(self, d):
+        """
+        Given a doc id, return the doc and its label bit vector.
+        """
+        return self._w_dp[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """
+        Given a list of labels, return the label bit vector.
+        An empty label list is an alias for all labels being set. This is
+        useful because otherwise when a new label is introduce to the model,
+        none of the input data needs to be altered.
+        """
+        if len(labels) == 0:
+            return numpy.ones(self._num_labels)
+        label_vector = numpy.zeros(self._num_labels)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        label_vector[self._label_to_id[self._DEFAULT_LABEL]] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, val):
+        """
+        Updates counting matrices (ones that begin with _c) when a label
+        is assigned and unassigned to a word.
+        """
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _increment_counting_matrices(self, d, w, l):
+        """
+        Assign a label to a word in a doc.
+        """
+        self._update_counting_matrices(d, w, l, 1)
+
+    def _decrement_counting_matrices(self, d, w, l):
+        """
+        Unassign a label from a word in a doc.
+        """
+        self._update_counting_matrices(d, w, l, -1)
+
+    def _get_doc_ids(self, doc_ids):
+        """
+        Helper function for methods that iterate over doc ids.
+        If no doc ids are provided, then all doc ids are iterated over.
+        This simplifies logic closer to the client.
+        """
+        if doc_ids is None:
+            doc_ids = xrange(self._num_docs)
+        return doc_ids
+
+    def _assign_initial_labels_to_word_instances(self, doc_ids=None):
+        """
+        Uniformly assigns labels to word instances in a new document. This
+        prevents the model from converging to a non-useful state when training.
+        """
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_doc_with_label_vector(d)
+            l_p = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
```

Comment:
I just realized something -- the comment I made at the top here might be the inspiration I was talking about in the previous pass, since I was uncomfortable about having this method (with its numerous prerequisites) be broken off from the add_examples() etc. methods).

In reality you have two different cases. You have a bunch of unlabeled docs (for which you pick a label from the set of all labels at random), and you have a bunch of labeled docs (for which you pick one of their assigned labels at random). These two cases are being combined into the same method, and I think that's giving rise to many of the comments.

How about just doing the update to self._l_dp inline, in the methods which already have the label lists? Then you don't have to worry about this being called when it shouldn't be, and all the initialization/updates to the internal variables (like _b_dl, etc.) happen in the same place. I get that this method looks a bit long at the moment, but I think that's in part due to all the passing stuff around. In addition, I suspect that the numpy stuff here can probably be simplified...